rosette-scala/src/main/scala/coop/rchain/rosette/Tuple.scala

```diff
@@ -1,14 +1,145 @@
 package coop.rchain.rosette
 
+/** Tuple
+  *
+  *  Tuples are used to represent messages that convey information
+  *  between two actors in a communication, to introduce sub-structure
+  *  in messages, and as a primitive data structure in the definition
+  *  of other actors
+  */
 case class Tuple(elem: Seq[Ob],
-                 override val entry: Seq[Ob],
+                 override val parent: Ob,
                  override val meta: Ob,
                  override val slot: Seq[Ob])
-    extends Ob
+    extends Ob {
+
+  def makeSlice(offset: Int, n: Int): Tuple =
+    this match {
+      case Tuple.NIL => Tuple.NIL
+      case _ => Tuple(n, this, offset, n)
+    }
+
+  def makeTail(entriesToSkip: Int): Tuple = {
+    val size = this.elem.size - entriesToSkip
+    Tuple(size, this, entriesToSkip, size)
+  }
+
+  def nth(n: Int): Option[Ob] = this.elem.lift(n)
+
+  def setNth(n: Int, ob: Ob): Option[Tuple] =
+    try {
+      Some(Tuple(this.elem.updated(n, ob), this.parent, this.meta, this.slot))
+    } catch {
+      case _: IndexOutOfBoundsException => None
+    }
+
+  def subObject(i: Int, n: Int): Tuple = makeSlice(i, n)
+
+  def accepts(msg: Ctxt): Boolean =
+    if (this == Tuple.NIL) {
+      true
+    } else {
+      this.elem.exists(_.matches(msg))
+    }
+
+  override def matches(msg: Ctxt): Boolean = {
+    val n = this.elem.size
+
+    if (n > 0 && n <= msg.nargs) {
+      if (this.elem.head != msg.trgt) {
+        false
+      } else {
+
+        /*
+         * Skip msg.head, since it is actually the receiver of the
+         * message, i.e.,
+         *
+         * 	(oprn rcvr arg1 arg2 ...)
+         *
+         * results in an argvec of
+         *
+         * 	[rcvr arg1 arg2 ...]
+         */
+
+        this.elem
+          .drop(1)
+          .zip(Stream from 0)
+          .forall {
+            case (e, i) =>
+              msg.arg(i) match {
```

Comment:
It seems odd to use `zip` above but then use `msg.arg(i)`; why not `zip` with `msg.arg` as well?

Reply:
You are right. I did it this way because of the case when `this.elem.size` > `msg.nargs`. However this won't happen because in line 48 we are checking against that.